--- bkp.c	2012-05-16 18:34:36.912345655 +1200
+++ buffalo.c	2012-05-16 18:36:30.858395023 +1200
@@ -895,6 +895,7 @@
 	int i;
 	int running=1; /* set to false to stop, FIXME make into a naughty global later */
 	char ch[7]; /* 6 is maximum len of utf8, 7th adds a nice \0 */
+	int alt=0; /* if esc was pressed then next character is treated is alt-key, corrects for faulty keyboards sending stray alt */
 
 	i_setup();
 
@@ -904,6 +905,18 @@
 	while( running ){
 		i_draw();
 		t_read(ch, 7);
+		if( ch[0] == 27 && ch[1] == 0 ){
+			/* either esc pressed or stray alt, treat next key as alt-key */
+			alt = 1;
+			continue;
+		}
+		if( alt ){
+			/* alt mode, need to insert alt */
+			alt = 0;
+			memmove( &ch[1], ch, 6 );
+			ch[0] = 27;
+		}
+
 		if( i_utf8len(ch) > 1 ){
 			cur = i_insert(cur, ch);
 		} else if( ch[0] == 127 ){
